torch/_functorch/_aot_autograd/runtime_wrappers.py

```diff
@@ -1868,24 +1878,46 @@ def get_types_for_tangents(tangents):
                         )
                         for i, t in enumerate(all_args)
                     ]
+                    # Coercing tangents memory format before unwrapping tensor subclasses,
+                    # As we have to coerce Subclass tangent first and then its Tensor attributes.
+                    assert (
+                        CompiledFunction.metadata.traced_tangent_memory_formats
+                        is not None
+                    )
+                    all_args = [
+                        (
+                            AOTDispatchAutograd.coerce_runtime_tangent_tracing_memory_format(
+                                t,
+                                CompiledFunction.metadata.traced_tangent_memory_formats[
+                                    i - tangents_start_idx
+                                ],
+                            )
+                            if tangents_start_idx <= i < tangents_end_idx
+                            else t
+                        )
+                        for i, t in enumerate(all_args)
+                    ]
                     all_args = unwrap_tensor_subclasses(
                         all_args, is_joint_structure=False
                     )
-                    tangents_start_idx = (
-                        len(all_args) - len_tangents - len(rng_args) - len(bw_tokens)
-                    )
-                    tangents_end_idx = tangents_start_idx + len_tangents
-
-                # Make the tangents contiguous. Note that we must do this after subclass desugaring
-                # because inputs to inductor have to be contiguous
-                all_args = [
-                    (
-                        AOTDispatchAutograd._force_contiguous(t)
-                        if (tangents_start_idx <= i < tangents_end_idx)
-                        else t
+                else:
+                    assert (
+                        CompiledFunction.metadata.traced_tangent_memory_formats
+                        is not None
                     )
-                    for i, t in enumerate(all_args)
-                ]
+                    all_args = [
+                        (
+                            AOTDispatchAutograd.coerce_runtime_tangent_tracing_memory_format(
```

Comment:
hmm why do we need to call `coerce_runtime_tangent_tracing_memory_format` separately on both sides of the branch above?

Reply:
For the subclasses case we want to call it before flattening. Potentially we can move coercing before subclasses branch I will try.